extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/output/RetryableS3OutputStream.java

```diff
@@ -199,15 +229,44 @@ private void pushCurrentChunk() throws IOException
   {
     currentChunk.close();
     final Chunk chunk = currentChunk;
-    try {
-      if (chunk.length() > 0) {
-        resultsSize += chunk.length();
+    if (chunk.length() > 0) {
+      s3UploadConfig.incrementCurrentNumChunks();
```

Comment:
It really doesn't sit well with me that we are able to modify a "config" class. Perhaps the nomenclature is off, however configs aren't supposed to be modified during the execution. They can be dynamically updated, however I'd still expect it to be stateless.

Reply:
+1, I think this information can be tracked in a singleton called `S3UploadManager` or something. It is also a little weird to inject the `ExecutorService` everywhere. That executor service can live inside `S3UploadManager` so that we have a single place to manager the lifecycle of the executor. Also, individual `RetryableS3OutputStream` would submit jobs to the `S3UploadManager`.
It would be this manager that is injected everywhere instead of the raw `ExecutorService`.

Reply:
Yeah, I couldn't decide a good name for it. I initially named it S3UploadHelper, but that seemed confusing as well as it doesn't have the upload logic. Hence I settled for S3UploadConfig, as it does define the configuration for the upload mechanism.
Please let me know if you have any better naming suggestions for this, definitely open to it!

Reply:
@kfaraz Have made the suggested change. Refactored S3UploadConfig + ExecutorService into S3UploadManager. Please have a look, thanks!